core: Make `Debug` impl of raw pointers print metadata if present

Enselic opened the pull request:
Make Rust pointers appear less magic by including metadata information in their `Debug` output.

This does not break Rust stability guarantees because `Debug` impl are explicitly exempted from stability:
https://doc.rust-lang.org/std/fmt/trait.Debug.html#stability

> ## Stability
>
> Derived `Debug` formats are not stable, and so may change with future Rust versions. Additionally, `Debug` implementations of types provided by the standard library (`std`, `core`, `alloc`, etc.) are not stable, and may also change with future Rust versions.

Note that a regression test is added as a separate commit to make it clear what impact the last commit has on the output.

Closes #128684 because the output of that code now becomes:

```
thread 'main' panicked at src/main.rs:5:5:
assertion `left == right` failed
  left: Pointer { addr: 0x7ffd45c6fc6b, metadata: 5 }
 right: Pointer { addr: 0x7ffd45c6fc6b, metadata: 3 }
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

Enselic committed:
tests: Add regression test for `Debug` impl of raw pointers

Enselic committed:
core: Make `Debug` impl of raw pointers print metadata if present

Make Rust pointers less magic by including metadata information in their
`Debug` output.

This does not break Rust stability guarantees because `Debug` output is
explicitly exempted from stability:
https://doc.rust-lang.org/std/fmt/trait.Debug.html#stability

Co-authored-by: Lukas <26522220+lukas-code@users.noreply.github.com>
Co-authored-by: Josh Stone <cuviper@gmail.com>

Noratrieb commented on tests/ui/fmt/ptr-metadata.rs:
```diff
@@ -0,0 +1,19 @@
+
+use std::fmt::Display;
+
+fn main() {
```
can you make this a core unit test instead?

Enselic replied:
Sure no problem, I'll fix whatever my reviewer(s) ask me to do.

Would you mind elaborating a bit on why though? Do you prefer it as a core  unit test because then the test can be run with a stage0 compiler, maybe? What I like about having it as a UI test is that `--bless` works when the output change.

Enselic replied:
Done. I had to (with the current test code) add "regex" as an allowed library dependency, but it is already present as an allowed rustc dependency, so it's fine I think.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12601829547/job/35123733023) [(plain)](https://github.com/rust-lang/rust/commit/b7951461ec1c52bae63417b4f54c7626d85c83c2/checks/35123733023/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#21 exporting to docker image format
#21 sending tarball 27.6s done
#21 DONE 40.2s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
   Compiling compiletest v0.0.0 (/checkout/src/tools/compiletest)
    Finished `release` profile [optimized] target(s) in 15.12s
##[endgroup]
##[group]Testing stage2 compiletest suite=ui mode=ui (x86_64-unknown-linux-gnu)
error: detected unknown compiletest test directive `normalize-stdout-test: "0x[0-9a-fA-F]+" -> "$$HEX"` in /checkout/tests/ui/fmt/ptr-metadata.rs:3
errors encountered during EarlyProps parsing: /checkout/tests/ui/fmt/ptr-metadata.rs
thread 'main' panicked at src/tools/compiletest/src/header.rs:68:13:
errors encountered during EarlyProps parsing
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Build completed unsuccessfully in 0:15:01
```

</details>

lukas-code commented on library/core/src/fmt/mod.rs:
```diff
@@ -2767,7 +2767,15 @@ impl Display for char {
+        if core::mem::size_of::<<T as core::ptr::Pointee>::Metadata>() == 0 {
+            pointer_fmt_inner(self.expose_provenance(), f)
+        } else {
+            f.write_str("Pointer { addr: ")?;
+            pointer_fmt_inner(self.expose_provenance(), f)?;
+            f.write_str(", metadata: ")?;
+            Debug::fmt(&core::ptr::metadata(*self), f)?;
+            f.write_str(" }")
```
```suggestion
            f.debug_struct("Pointer")
                .field_with("addr", |f| pointer_fmt_inner(self.expose_provenance(), f))
                .field("metadata", &core::ptr::metadata(*self))
                .finish()
```

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12603163581/job/35127682970) [(plain)](https://github.com/rust-lang/rust/commit/ee8187af5b695da19bdf5fecb92565b0f504e796/checks/35127682970/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
---- fmt::test_pointer_formats_data_pointer stdout ----

thread 'fmt::test_pointer_formats_data_pointer' panicked at library/core/tests/fmt/mod.rs:18:5:
assertion `left == right` failed
  left: "Pointer { addr: 0x1, metadata: 0 }"
 right: "0x1"

failures:
    fmt::test_pointer_formats_data_pointer

```

</details>

Enselic committed:
tests: Use `as *const _` instead of `.as_ptr()` in ptr fmt test

Because `.as_ptr()` changes the type of the pointer (e.g. `&[u8]`
becomes `*const u8` instead of `*const [u8]`), and it can't be expected
that different types will be formatted the same way.

bors commented:
:umbrella: The latest upstream changes (presumably #135937) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Enselic commented:
Amanieu seems busy. Re-rolling reviewer.

r? libs

thomcc commented:
This needs an FCP.

@rustbot fcp merge

thomcc commented:
Oh right, wrong bot.

@rfcbot fcp merge

Amanieu commented on library/core/src/fmt/mod.rs:
```diff
@@ -2776,7 +2776,14 @@ impl Display for char {
 impl<T: ?Sized> Pointer for *const T {
     fn fmt(&self, f: &mut Formatter<'_>) -> Result {
-        pointer_fmt_inner(self.expose_provenance(), f)
+        if core::mem::size_of::<<T as core::ptr::Pointee>::Metadata>() == 0 {
```
It seems a bit strange to check the size of the metadata. Could we instead explicitly check that the metadata type is `()`?

Enselic replied:
I tried doing that at first but couldn't figure out how. I tried a bit more now but without success. I can't find any other place in the compiler/libraries that does that either.

Maybe you had some particular technique in mind?

Personally I get peace of mind thinking of this logic as "if the metadata has no size there is nothing of interest to show".

cuviper replied:
If we added a `'static` requirement to `Pointee::Metadata`, then you could compare `TypeId`. I'm not sure if there's any good reason it doesn't have that already.

Enselic replied:
Thanks, that seems to work. I think I can push a commit for that to see if CI passes without disrupting the FCP?

lukas-code replied:
`Pointee::Metadata` can't be `'static`, because the metadata of `dyn Trait + 'a` is `DynMetadata<dyn Trait + 'a>` which is not `'static`.

To me it seems safer to drop the `'static` bound from the `type_id` intrinsic rather than potentially causing a trait system unsoundness by adding a `'static` bound that isn't actually satisfied by the compiler-builtin impls.

Enselic replied:
That makes sense. I'll let CI continue to see if it catches it, and then go back to the code I had when the FCP started.

cuviper replied:
You could use a local specialization:

```rust
trait IsUnit {
    fn is_unit() -> bool;
}

impl<T: ?Sized> IsUnit for T {
    #[inline(always)]
    default fn is_unit() -> bool { false }
}

impl IsUnit for () {
    #[inline(always)]
    fn is_unit() -> bool { true }
}
```

(It would be better as an associated const, but I don't think `min_specialization` allows that.)

Enselic replied:
Thanks. I have now
* rebased on master to resolve conflict
* added a comment for future adventurers explaining why `Pointee::Metadata` can't be `'static`
* check for `()` with `IsUnit` instead of `size_of() == 0`

Enselic commented on library/core/src/ptr/metadata.rs:
```diff
@@ -61,7 +61,7 @@ pub trait Pointee {
     // NOTE: Keep trait bounds in `static_assert_expected_bounds_for_metadata`
```
and now I see this note to keep things in sync....

Enselic committed:
core: Document why Pointee::Metadata can't have 'static bound

Co-authored-by: Lukas <26522220+lukas-code@users.noreply.github.com>

bors commented:
:umbrella: The latest upstream changes (presumably #137046) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

thomcc commented:
Sorry, forgot to merge this.

@bors r+

bors commented:
:pushpin: Commit 697737a8b18b685270f197ccdd1211e8a010b98b has been approved by `thomcc`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=thomcc 697737a8b18b685270f197ccdd1211e8a010b98b -->
<!-- homu: {"type":"Approved","sha":"697737a8b18b685270f197ccdd1211e8a010b98b","approver":"thomcc","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
